cmd/api/insert.go

```diff
@@ -54,43 +45,71 @@ func (a *API) Insert(c *fiber.Ctx) error {
 			Msg("Incoming request")
 	}
 
+	body := c.Body()
+	if !gjson.ValidBytes(body) {
+		return fiber.NewError(http.StatusBadRequest, "invalid JSON")
+	}
+
 	// TODO: this block can be abstracted as we also use it for query
-	/////
 	apiKey := c.Locals("apiKey").(models.APIKey)
 
 	// TODO: read-only vs read-write connections
 	connectionSetting := a.db.GetDatabaseConnection(apiKey.DestinationID)
-
 	if connectionSetting.ID == "" {
-		return errors.New("No DB Connections set up")
+		return fiber.NewError(http.StatusUnauthorized, "no connection is set up")
 	}
-	/////
 
-	input := c.Body()
-	if !gjson.ValidBytes(input) {
-		return errors.New("Invalid JSON")
+	var (
```

Comment:
Is there a way we can simplify the logic in this function? I realize it is complicated: we might get parameters from 3 different places (header, query param, request body) and data could be a list or an object. I would love for this function to read more linearly so we can better understand which path a piece of data is taking through the system.

Reply:
There is/was an existing `getFlattenType` method that I think did the right thing so I think it just need generalizing. Something like https://github.com/sd1li/ScratchDB/commit/fce694fee110645e273faed94fc74e019c1a34af . the names might need improving, just the gist is to have a dedicated type `DataKeys` that looks up the relevant value in the headers/query/body. I also used a n enum `DataSource` instead if "strings" for clarity.